### PR TITLE
Full Height Alignment Toolbar: Implementation + Cover integration.

### LINKED ISSUE #16385: Full Screen alignment/display option on several blocks
**Is your feature request related to a problem? Please describe.**
Would be really nice to have ability to set several blocks as full screen. It's a very powerful option to have on Cover and Media Text Blocks.

Tagging @mtias :)

**Describe the solution you'd like**
I've added this feature on [EditorsKit](https://wordpress.org/plugins/block-options/) plugin and here's the preview : https://twitter.com/phpbits/status/1143122996552335362

### PR DESCRIPTION
**Pull request by retrofox**
## Description
This PR adds the Full Height Toolbar Alignment component to the block editor package, as well as its implementation to ~`Media & Text` and~ `Cover` blocks.

I did my first attempt to achieve this a long time ago [in this PR](https://github.com/WordPress/gutenberg/pull/16738). Although I couldn't finish it, it was very good to get feedback. I collected all ideas and implemented a solution (mostly a suggestion actually) to try to tackle all requirements. Spoiler Alter: Alignments never was for beginners.

Fixes #16385

Previous PRs https://github.com/WordPress/gutenberg/pull/16738 https://github.com/WordPress/gutenberg/pull/19767

## Approach

Unlike the block alignment options, `Full Height Alignment` can be thought of as an independent alignment type. It can be applied independently of other alignments, but also, it works very very well together with these block alignment options, where the combination empowers the design-layout capability.

The buttons disposition in the toolbar could be something like the following:

image

Above, it shows the _**toolbar alignment options**_ next to the _**full height alignment option**_.

~_Note: I did the icon, 100% opened to change it_~

## Integrating with Cover block

The Cover block is a lovely case to see in action because this block also has implemented the min-height control. So the goal here is integrating this control with the Full Height Alignment. I'd like to say it was my idea, [but it wasn't](https://github.com/WordPress/gutenberg/issues/16385#issuecomment-640588650). Thanks, @mtias 

In the following example, the block has `full width`, `full height`, content position `left bottom` and `fixed background`.

Screen Shot 2020-11-04 at 7 13 10 PM

The result is very nice:

full-height-cover-7

### **Full Height Alignment** and **Dimensions control**.

I've connected both components to be consistent in the way that it applies the height to the cover block. In this case, Full Height Aligents acts just as a shorthand to apply `height` to `100vh`. You can at any time change or those values, and the button gets "untoggled" once that happens. In fact, It doesn't add any new attribute to the block.

From Button to Control | From Control to Button
-|-
full-height-cover-10 | full-height-cover-11

## How has this been tested?

* Apply these changes
* Play with the components
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by jasmussen** [comment]
Thank you for working on this. This is what I see:

full

Since we last discussed this feature, I've had a chance to think a bit more about how it might behave. Specifically, we might want to have the toolbar button for just a few blocks, but instead of applying a classname (`is-full-height-aligned`), we might want to simply set an inline style, like you can with Cover:

cover

The reason being that height is a little different from width in a web that is mostly a vertical flow of content. For width, it might make sense to have specific width presets (and indeed allow these on all block, #25973), which could look like this in the sidebar:

Screenshot 2020-11-02 at 09 25 19

But height should probably remain flexible and fluid, and have an input field. It could probably still have a toolbar button, like this PR adds, but it should only be applied to a few blocks. But the design is a bit tricky. Here's an attempt at a better icon:



SVG:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M5 9h1.5V6.5H9V5H5v4zm1.5 6H5v4h4v-1.5H6.5V15zM15 5v1.5h2.5V9H19V5h-4zm2.5 12.5H15V19h4v-4h-1.5v2.5z" fill-rule="evenodd" clip-rule="evenodd" fill="#1e1e1e"/></svg>
```

I know the removal of the classname is a bit of a change from the initial discussions. I'd be curious for thoughts by @jameskoster, @ItsJonQ and @mtias

**Comment by jameskoster** [comment]
Having a toolbar button to quickly set the height to 100vh seems like a handy shortcut, but I would agree it's probably best restricted to just a few blocks.

It may be something to discuss in another issue, but I find the alignment icons themselves quite tricky to understand, particularly wide and full width. If we're going to add an icon for full height, I wonder if we might revisit and attempt to find consistency across all these alignment icons?

**Comment by retrofox** [comment]
Thanks, @jasmussen for your review

> Thank you for working on this. This is what I see:

Sorry, do you see something wrong there? Just in case, the full height is applied to the block, not to the image component.

1 | 2 | 3
-|-|-
image | image | image

If you'd like to stretch full height, you should enable `Crop image to fill entire column`. This is an expected behavior, correct? 

> Specifically, we might want to have the toolbar button for just a few blocks...

This PR adds `Full Height` only to `Media & Text` and `Cover` blocks.

>..., but instead of applying a classname (is-full-height-aligned), we might want to simply set an inline style, like you can with Cover:

Sounds good to me. Going to work on this.

> For width, it might make sense to have specific width presets (and indeed allow these on all block, #25973), which could look like this in the sidebar:

Looks great to me. Just let's keep in mind that I think it worths refactoring the current implementation before starting to iterate over this since if I'm not wrong, this component is defined in the `core/cover` component. We should expose it to make it easily usable.

> But height should probably remain flexible and fluid, and have an input field.

👍 

> It could probably still have a toolbar button, like this PR adds, but it should only be applied to a few blocks.

It's added to only two blocks :-D

> But the design is a bit tricky. Here's an attempt at a better icon:

Excellent.

**Comment by retrofox** [comment]
Thanks, James.

> Having a toolbar button to quickly set the height to 100vh seems like a handy shortcut, but I would agree it's probably best restricted to just a few blocks.

Yes, agree :-D


> It may be something to discuss in another issue, but I find the alignment icons themselves quite tricky to understand, particularly wide and full width. If we're going to add an icon for full height, I wonder if we might revisit and attempt to find consistency across all these alignment icons?

The reason why I've put it in a different section of the toolbar is that the most powerful feature of this ( seemingly simple) alignment is the combination with other alignment types.

Just FYI, we've discussed a lot in the [previous PR](https://github.com/WordPress/gutenberg/pull/16738).

**Comment by jasmussen** [comment]
@jameskoster 

> It may be something to discuss in another issue, but I find the alignment icons themselves quite tricky to understand, particularly wide and full width. If we're going to add an icon for full height, I wonder if we might revisit and attempt to find consistency across all these alignment icons?

Absolutely. I was having the same thought notably with the Media & Text toolbar, which becomes very abstract with the black bar icons.

> Sorry, do you see something wrong there? Just in case, the full height is applied to the block, not to the image component.

No no, the feature is working fine! The GIF was perhaps misleading because the image was too low resolution to fill the vertical height.

> Looks great to me. Just let's keep in mind that I think it worths refactoring the current implementation before starting to iterate over this since if I'm not wrong, this component is defined in the core/cover component. We should expose it to make it easily usable.

Yes indeed. For starters, though, I think the approach to agree on is whether we should remove the full-height classname and replace it with an inline style. I think that could be a good first step to unifying the dimensions.

**Comment by retrofox** [comment]
> Yes indeed. For starters, though, I think the approach to agree on is whether we should remove the full-height classname and replace it with an inline style. I think that could be a good first step to unifying the dimensions.

It already applies the height using inline styles in the `cover` block. I was thinking if we could remove the `Media & Text` part, and iterate in another follow-up PR for this block.

**Comment by retrofox** [comment]
Updated icon here e73765ceee6b34f20ac4949c030369ae76222b5a

before | after
-|-
image | image

**Comment by retrofox** [comment]
> Since we last discussed this feature, I've had a chance to think a bit more about how it might behave. Specifically, we might want to have the toolbar button for just a few blocks, but instead of applying a classname (is-full-height-aligned), we might want to simply set an inline style, like you can with Cover:

Just to be clear, it applies inline style right not in this PR :-D

**Comment by jasmussen** [comment]
> Just to be clear, it applies inline style right now in this PR :-D. I've written a section about this in the PR description.

Thank you for the clarification, definitely worth looking closely at the bottom of your PR description 🙈  — I blame this week being unusually exhausting.

What I'm saying though, is that we need to not have the CSS class at all:

Screenshot 2020-11-02 at 15 05 36

The way I see it, this should only be two controls:

- A height input field with optional vh unit
- A toolbar toggle

Those two could/should play together as you do now, insofar as if the height is set to 100vh, the button is toggled. Is that possible to do?

**Comment by retrofox** [comment]
> What I'm saying though, is that we need to not have the CSS class at all:

Let me clean the `Media & Text` part, for now, from this PR. I'd like to keep working with the cover because this block has the min-height unit control implemented. We need to be consistent between both components.



> > Just to be clear, it applies inline style right now in this PR :-D. I've written a section about this in the PR description.
> 
> Thank you for the clarification, definitely worth looking closely at the bottom of your PR description 🙈 — I blame this week being unusually exhausting.

:-) No worries.

> 
> What I'm saying though, is that we need to not have the CSS class at all:
> 
> Screenshot 2020-11-02 at 15 05 36

> What I'm saying though, is that we need to not have the CSS class at all:

Let me clean the `Media & Text` part (it handles the heigh via CSS class), for now, from this PR. I'd like to keep working with the cover because this block has the min-height unit control implemented. We need to be consistent between both components (full height and unit controls)


> The way I see it, this should only be two controls:
> 
> * A height input field with optional vh unit

Already implemented in the cover component, correct?

image


> * A toolbar toggle

Already implemented in this PR.

> Those two could/should play together as you do now, insofar as if the height is set to 100vh, the button is toggled. Is that possible to do?

It is at all! and this is something that I did before but... (always there is "but") there is a minor issue here that made me add the toggle in the sidebar (I tried to explain it in the PR too 😅)

> It worths mentioning that at least for the cover block and in this implementation, the **_full height alignment_** is slightly different than just applying _**min-height: 100vh**_, basically because we want to guarantee the block doesn't stretch more than the screen. In behind, it also applies height: 100vh :smart:

does it make sense of this concern for you?

**Comment by retrofox** [comment]
> * A height input field with optional vh unit

Keep in mind that unit control applies `min-height` to the cover, instead of `height`.

image

Just saying because this small difference could be an issue if we aren't totally aware.

**Comment by jasmussen** [comment]
Yep, that's a good differentiation, thank you for the clarification. It does seem like we'll only ever want one field for height, if we can do with it, whether that's min-height or plain height. I think I understand the problem, and I also think I need to give my brain a rest to think more about it 😅

The difficult part here is getting the interface right!

**Comment by retrofox** [comment]
In terms of functionality, if we don't apply `hight` to the block, it will stretch if the content forces it.

Screen Shot 2020-11-02 at 12 39 36 PM

Applying `height` to the block, it will try to keep this value to `100vh`, independently of its content. In other words, it will follow the viewport height (green arrow). If we apply `min-height: 100vh`, the block height will stretch wrapping the content.

**Comment by retrofox** [comment]
**Toggle `100vh` by clicking on `Full Height` toolbar button**

full-height-cover-10

**Toggle active state of `Full Height` toolbar button, setting the min-height from block sidebar.**

full-height-cover-11

**Comment by retrofox** [comment]
> The way I see it, this should only be two controls:
> 
> * A height input field with optional vh unit
> * A toolbar toggle
> 
> Those two could/should play together as you do now, insofar as if the height is set to 100vh, the button is toggled. Is that possible to do?

Addressed [here](https://github.com/WordPress/gutenberg/pull/26615/commits/510c0fdafdb0eec6509cc5fd82c2de84d5328dc5), and also added some gifs [here](https://github.com/WordPress/gutenberg/pull/26615#issuecomment-720620277).

**Comment by jasmussen** [comment]
This is getting there:

current state

I would say the user experience is now very close:

- The fullscreen button is a shortcut to set 100vh in the sidebar
- You can at any time change or those values, and the button gets "untoggled" once that happens

Markup also looks good:

Screenshot 2020-11-03 at 10 09 18

This is growing on me, but because it touches both the cover block and the min-height values, this needs a little sanity checking. Here are a few that have worked on related properties: @sirreal @nosolosw @jorgefilipecosta — thanks for checking!

Visually, I would like to see a small change to the toolbar as it looks here:

Screenshot 2020-11-03 at 10 08 11

Can we do this instead:

Screenshot 2020-11-03 at 10 08 11

That is — the fullscreen button feels like it belongs with the other alignments.

**Comment by retrofox** [comment]
> Can we do this instead:
> 
> Screenshot 2020-11-03 at 10 08 11

Very nice idea. Going to work on this. I think we'll have to move the toolbar and out all buttons together in the _block alignments toolbar_

**Comment by retrofox** [comment]
> Can we do this instead:

It's going quite complex, and the needed changes to make it work will make the current approach quite more complex. That's something that we want to do?

Personally, I think it should be discussed in a separate PR and deserve its own space. Wondering if we can get with the current approach and then continue with a new follow-up PR.

Anyway, I've been working on this integration at least for researching purposes.

**Comment by paaljoachim** [comment]
Thank you for working on this again Damien @retrofox !

The Cover block Full Height setting works nicely.

For the Media & Tex block. This is what I see (I will repeat some of the same things that are mentioned above).
From the perspective of how I experience them.

Clicking the Full Height icon the image jumps downward (similar to what Joen mentioned above).
Screen Shot 2020-11-03 at 18 34 30

I expect when clicking the Full Height icon that the image is seen in full height. Which means that the image is (soft) cropped to fit in the view port area. Perhaps we need a focal point picker similar to the Cover block to where one ends up moving the image to show the portion of the image in the view port one wants to see. 

I do understand that clicking the Full Height icon brings the full block into a full height alignment, but this is not easily seen.
So right away I might think that clicking the icon is not working because the image did not go into full height.

**Comment by retrofox** [comment]
> Thank you for working on this again Damian @retrofox !
✋ 


> Clicking the Full Height icon the image jumps downward (similar to what Joen mentioned above).

> I expect when clicking the Full Height icon that the image is seen in full height. Which means that the image is (soft) cropped to fit in the view port area. Perhaps we need a focal point picker similar to the Cover block to where one ends up moving the image to show the portion of the image in the view port one wants to see.

I think it's already implemented.

> I do understand that clicking the Full Height icon brings the full block into a full height alignment, but this is not easily seen.
> So right away I might think that clicking the icon is not working because the image did not go into full height.

It's a fair enough point of view, really. I answered the technical [question here](https://github.com/WordPress/gutenberg/pull/26615#issuecomment-720430176). While I develop this feature I realize that it could behave not totally equal for different blocks. `Media&Text` is an example.
As I said before, the full height is applied to the whole block, not for the image. But we can think that `Full Height` for the `Media & Text` blocks implies two things:

* set the block height to the same value as the viewport height
* Stretch the inner image to block height. It means to activate the "Crop Image to Fill the entire column" option

image

**Comment by jasmussen** [comment]
Thanks for keeping on this retrofox!

> It's going quite complex, and the needed changes to make it work will make the current approach quite more complex. That's something that we want to do?

It's something we want to do across blocks, here for instance it's tracked for the Image block: #26411. The question is whether doing that should block this PR, or whether it's fine to proceed.

Specifically, the Media & Text block is very complicated rigth now, with icons that need a bit of love. Given there are also some full-height inconsistencies (as you've explained already), maybe we should remove the button from Media & Text for now, and keep it only in Cover? The main Cover toolbar is simple enough to carry the extra button. 

Screenshot 2020-11-04 at 09 44 46

I looks like you removed the border separator between alignment and full height, but in the markup I still see two toolbar groups, so the spacing is wrong. Unless we can group the buttons together in the same group (to your point, maybe something we can look at separately) then we shouldn't remove the border. 

So to summarize, it seems like we could try this as a minimal first version:

- Keep the border between align and full height (unless they can be in the same toolbar group)
- Don't add any CSS classes, only the inline style that's the same as set by the sidebar control
- Add the button only to Cover

Addressing those makes for a small PR, but it might be a great way to try the shortcut, and then we can see whether it's useful or not. What do you think?

**Comment by retrofox** [comment]
> Thanks for keeping on this retrofox!

✋ 

> It's something we want to do across blocks, here for instance it's tracked for the Image block: #26411. The question is whether doing that should block this PR, or whether it's fine to proceed.

Agree. I really do think of grouping toolbars is something that will make the interfacer clearer to consume. What you pointed is a good example. This implementation opens the room for other questions, that I think we should keep in the issue.

> Specifically, the Media & Text block is very complicated rigth now, with icons that need a bit of love. Given there are also some full-height inconsistencies (as you've explained already), maybe we should remove the button from Media & Text for now, and keep it only in Cover? The main Cover toolbar is simple enough to carry the extra button.

Yes... `Media & Text` is a block as tricky as a powerful IMO, and both things are tied, naturally. For instance, stretch the image height when the block gets the full height alignment, automatically on the fly (_Crop Image to Fill the entire column_), to get something that the following.



> I looks like you removed the border separator between alignment and full height, but in the markup I still see two toolbar groups, so the spacing is wrong.

Yes, it was kind of intended. They are two toolbars :-/. Something to discuss https://github.com/WordPress/gutenberg/issues/26411, too.

> Unless we can group the buttons together in the same group (to your point, maybe something we can look at separately) then we shouldn't remove the border.

Sounds like the faster and light option to take.

> So to summarize, it seems like we could try this as a minimal first version:

> * Keep the border between align and full height (unless they can be in the same toolbar group)
> * Don't add any CSS classes, only the inline style that's the same as set by the sidebar control
> * Add the button only to Cover

> Addressing those makes for a small PR, but it might be a great way to try the shortcut, and then we can see whether it's useful or not. What do you think?

100% agree. Exciting.

**Comment by retrofox** [comment]
- [x] Keep the border between align and full height 
- [x] Don't add any CSS classes, only the inline style that's the same as set by the sidebar control
- [x] Add the button only to Cover

RfR again.
I'm afraid you'll have to get a fresh pull from this branch because I've rebased it. Thanks in advance.

**Comment by jasmussen** [comment]
Nice work. This is what I see:

fullheight

This now works as I would expect such a button to work. 

_Very minor nitpick, instead of the button tooltip saying "Full Height", can it say "Toggle full height"?_

Other than that, I think this is good enough to test it, 👍 👍. All it needs now is a code review. @ItsJonQ do you have the energy to do so?

**Comment by retrofox** [comment]
> Very minor nitpick, instead of the button tooltip saying "Full Height", can it say "Toggle full height"?

you know, I had a suggestion, which in fact it was implemented here but then removed because ... who knows, that tweaked the label of the button. Let me restore it. Sounds good to me. Thanks for your help.

**Comment by retrofox** [comment]
I've updated the label [here](https://github.com/WordPress/gutenberg/pull/26615/commits/fe64be8ca8056b445befa90cb91c136a7a5e94cb) but also implemented [another idea](https://github.com/WordPress/gutenberg/pull/26615/commits/f0b85e1354b28088069b3fa25cdcb0eca953b44f). Please feel free to discard it.

It changes the button label depending on the height values.

**Set full height**


**Unset full height**
It restores to the default height.



**Toggle to previous height**

**Comment by mapk** [comment]
@retrofox, I've been watching this PR and excited to see it get in! ❤️  

I prefer "Toggle full height" to those other options personally. "Toggle" seems to align nicely with the "Change" verb that is used for the other toolbar icons. And with "toggle" you don't need to specify the on/off state with words.

**Comment by retrofox** [comment]
> @retrofox, I've been watching this PR and excited to see it get in! ❤️
> 
> I prefer "Toggle full height" to those other options personally. "Toggle" seems to align nicely with the "Change" verb that is used for the other toolbar icons. And with "toggle" you don't need to specify the on/off state with words.

yeah... I agree honestly, but I liked to play with the dynamic label. Removed ✔️

File: packages/block-editor/src/components/block-full-height-alignment-toolbar/icon.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/**
+ * WordPress dependencies
+ */
+import { Path, SVG } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+
+export default (
```
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/block-full-height-alignment-toolbar/icon.js:10]: We have a WordPress/icons package. Would it make sense to add this icon there?

File: packages/block-library/src/cover/edit.js [hunk 2]
```diff
@@ -264,6 +266,47 @@ function CoverEdit( {
...
 
+	const [ prevMinHeightValue, setPrevMinHeightValue ] = useState( minHeight );
+	const [ prevMinHeightUnit, setPrevMinHeightUnit ] = useState(
+		minHeightUnit
+	);
+	const isMinFullHeight = minHeightUnit === 'vh' && minHeight === 100;
+
+	// Remember previous values of height,
+	// to be able to implement Full Height toggle.
+	useEffect( () => {
```
Comment (jorgefilipecosta) [inline packages/block-library/src/cover/edit.js:277]: Would it work to each time we execute toggleMinFullHeight, and isMinFullHeight was not true we store the previous value inside toggleMinFullHeight removing the need for an effect ?

File: packages/block-library/src/cover/edit.js [hunk 3]
```diff
@@ -264,6 +266,47 @@ function CoverEdit( {
 	const onSelectMedia = attributesFromMedia( setAttributes );
 	const isBlogUrl = isBlobURL( url );
 
+	const [ prevMinHeightValue, setPrevMinHeightValue ] = useState( minHeight );
```
Comment (jorgefilipecosta) [inline packages/block-library/src/cover/edit.js:269]: I guess these two state values could share the same useState call the value being an object with properties minHeight and minHeightUnit that could be directly passed to set attributes. These change also aligns with probable change we will have to do make minHeightUnit and minHeight a single attribute (a string that can contain multiple CSS valid properties).

If we don't initiate the state of previous to minHeight, minHeightUnit e.g: keep them undefined and also follow the suggestion of setting the previous values inside toggleMinFullHeight we can also probably remove the need for this condition bellow "// If there aren't previous values, take the default ones.".

**Review by jorgefilipecosta — APPROVED** [review]
I left some suggestions that may simplify the code but from the functional point of view it works well 👍

File: packages/block-editor/src/components/block-full-height-alignment-toolbar/icon.js (hunk 1 above)
Comment (jasmussen) [reply packages/block-editor/src/components/block-full-height-alignment-toolbar/icon.js:10]: I'd say so, yes.

File: packages/block-library/src/cover/edit.js (hunk 3 above)
Comment (retrofox) [reply packages/block-library/src/cover/edit.js:269]: Nice suggestion. Going to refactoring using an object.

> These change also aligns with probable change we will have to do make minHeightUnit and minHeight a single attribute (a string that can contain multiple CSS valid properties).

💯 agree.
I did this approach [in the Background Size implementation](https://github.com/WordPress/gutenberg/pull/26976). The data is stored in the [backgroundSize attribute (string)](https://github.com/WordPress/gutenberg/blob/bf596cc512f411b9b2d8991db2f3499a537c7e66/packages/block-library/src/cover/block.json#L20-L23)ready to be used in the [inline style](https://github.com/WordPress/gutenberg/blob/bf596cc512f411b9b2d8991db2f3499a537c7e66/packages/block-library/src/cover/edit.js#L486). The block takes over of parsing and converting from/to this attribute to get the data usable by the UI components.

Happy to do this in a follow-up PR.

File: packages/block-library/src/cover/edit.js (hunk 2 above)
Comment (retrofox) [reply packages/block-library/src/cover/edit.js:277]: let's try it Sounds smart to me :-)
